crates/engine/tree/src/tree/mod.rs

```diff
@@ -120,6 +134,11 @@ impl TreeState {
             }
         }
     }
+
+    /// Returns the maximum block number stored.
+    pub(crate) fn max_block_number(&self) -> BlockNumber {
+        *self.blocks_by_number.last_key_value().unwrap_or((&BlockNumber::default(), &vec![])).0
```

Comment:
Similar approach as in `get_blocks_to_persist`, assuming `blocks_by_number` is the source of truth for blocks to write to db, let me know if that's ok